lib/change_stream.js

```diff
@@ -343,58 +344,79 @@ var buildChangeStreamAggregationCommand = function(self) {
 };
 
 // Handle new change events. This method brings together the routes from the callback, event emitter, and promise ways of using ChangeStream.
-var processNewChange = function(self, err, change, callback) {
+var processNewChange = function(self, err, change, callback, isStream) {
   // Handle errors
   if (err) {
+    if (!isResumableError(err)) {
+      if (isStream) {
+        return self.emit('error', err);
+      }
+
+      else if (typeof callback === 'function' ) {
+        return callback(err, null);
+      }
+      else {
+        return self.promiseLibrary.reject(err);
+      }
+    }
+
     // Handle resumable MongoNetworkErrors
-    if (isResumableError(err) && !self.attemptingResume) {
+    if (isResumableError(err)) {
       self.attemptingResume = true;
 
       if (!(getResumeToken(self) || getStartAtOperationTime(self))) {
         const startAtOperationTime = self.cursor.cursorState.operationTime;
         self.options = Object.assign({ startAtOperationTime }, self.options);
       }
 
-      if (callback) {
-        return self.cursor.close(function(closeErr) {
-          if (closeErr) {
-            return callback(err, null);
-          }
-
-          self.cursor = createChangeStreamCursor(self);
+      if (isStream) {
+        self.cursor = createChangeStreamCursor(self);
```

Comment:
Now an `aggregate` command is sent after the `killCursors` (when `rs.stepDown()` is run on the primary). This `aggregate` fails with a `notMaster` error, which happens during two situations:
1) the `aggregate` runs before the primary steps down and the `replicaSetState` shows that it is still the primary
2) the `aggregate` runs before a new primary steps up and the `replicaSetState` shows that the old primary's state is undefined

The solution should be to first wait until a new primary is elected and to then run the `aggregate` command on the new primary. I'm not sure how to force the resume to wait until a new primary is elected so I'd appreciate any insight @mbroadst .